src/sampling.jl

```diff
@@ -42,6 +42,47 @@ function direct_sample!(rng::AbstractRNG, a::AbstractArray, x::AbstractArray)
 end
 direct_sample!(a::AbstractArray, x::AbstractArray) = direct_sample!(Random.GLOBAL_RNG, a, x)
 
+# check whether we can use T to store indices 1:n exactly
+for T in unique!([Int,UInt,Int64,UInt64,Int128,UInt128])
+    @eval isexact(n, ::Type{$T}) = true
+end
+isexact(n, ::Type{T}) where {T<:Integer} = n ≤ typemax(T)
+isexact(n, ::Type{BigInt}) = false # bigint is too expensive
+for T in (Float32, Float64)
+    @eval isexact(n, ::Type{$T}) = n ≤ maxintfloat($T)
+end
+isexact(n, ::Type{Complex{T}}) where {T} = isexact(n, T)
+isexact(n, ::Type{Rational{T}}) where {T} = isexact(n, T)
+isexact(n, T) = false
+
+# order results of a sampler that does not order automatically
+function sample_ordered!(sampler!::F, rng::AbstractRNG, a::AbstractArray, x::AbstractArray) where {F}
+    n, k = length(a), length(x)
+    if isexact(n, eltype(x))
+        sort!(sampler!(rng, Base.OneTo(n), x), by=real)
+        for i = 1:k
```

Comment:
Maybe worth it?
```suggestion
        @inbounds for i = 1:k
```

Reply:
Sure.   It worries me a little that the code never checks for 1-based indices, however.  (But this applies to existing code too.)

Reply:
Yeah, it would be a good idea to add an assertion somewhere.

Reply:
Yeah, it would be a good idea to add an assertion somewhere.